### PR TITLE
Stock Service Updates

### LINKED ISSUE #191: Moving products between locations, i.e. freezer to fridge
What's the best way to handle this scenario? You go to the butcher and buy a chunk of meat that's planned for in a few weeks. It initially goes into the freezer with a long expiry date. However, when you pull it out to defrost, it moves to the fridge and the expiry date needs to change. 

You can choose the location of a product when you purchase it. However, that location and the expiration date are then uneditable.  Could the fix be as simple as making the product editable?

### LINKED ISSUE #313: Option to edit current price in masterdata
It would be very nice to haven an option to edit the current price of a product in master data.

### PR DESCRIPTION
**Pull request by kriddles**
Closes #191 
Closes #313

### CODE REVIEW

**Comment by berrnd** [comment]
Wow, thanks a lot! :)

Will review and integrate this soon, but maybe could take a week or so...

**Comment by kriddles** [comment]
:) you're welcome

Yes, take your time

**Comment by kriddles** [comment]
Thank you for reviewing my suggested changes. 
I have made a few changes to keep the stock_id aligned with your intent which I fixedup in the AddProduct commit.

I think it's best for me to provide an example to explain my thoughts on stock_id and my recent updates.
If I buy 2 bottles of milk on Monday then I will have a 

Row 1
- stock_id = 1a1a
- purchased_date = 10/12/2019
- best_before_date = 11/12/2019
- amount = 2
- price = 2.50
- location_id = 1 (main refrigerator)

If I go back to the store the same day because I need 1 more bottle of milk and purchase it at the same price would you want a new stock_id? I would like to instead see the amount updated for stock_id 1a1a to be 3 since the purchased_date, best_before_date, price, and location_id would all be the same.

Row 1
- stock_id = 1a1a
- purchased_date = 10/12/2019
- best_before_date = 11/12/2019
- amount = 3
- price = 2.50
- location_id = 1 (main refrigerator)

This is the same concept for now adding the transfer feature.
Moving 1 bottle of milk from the location 1 (main refrigerator) to a location 2 (garage refrigerator) then I would not expect a new stock_id. The only difference would be a new row with a different location_id and amount would be updated from the initial location 1 so that the 2 rows have 1 amount.

Row 1
- stock_id = 1a1a
- purchased_date = 10/12/2019
- best_before_date = 11/12/2019
- amount = 2
- price = 2.50
- location_id = 1 (main refrigerator)

Row 2
- stock_id = 1a1a
- purchased_date = 10/12/2019
- best_before_date = 11/12/2019
- amount = 1
- price = 2.50
- location_id = 2 (garage refrigerator)

If I decide to transfer the bottle back to location_id 1 then I would see the row deleted for location_id 2 and see the initial row updated to amount 3 again.

Row 1
- stock_id = 1a1a
- purchased_date = 10/12/2019
- best_before_date = 11/12/2019
- amount = 3
- price = 2.50
- location_id = 1 (main refrigerator)
Row 2
- deleted

Let's say I go back to the store the following day and buy 1 more bottle of milk. In this case, I would be creating a new stock_id since the purchased_date is different even though the best_before_date and price maybe the same.

Row 3
- stock_id = 2b2b
- purchased_date = 10/13/2019
- best_before_date = 11/12/2019
- amount = 1
- price = 2.50
- location_id = 1 (main refrigerator)

This is what I attempted to code in the AddProduct to check for stock that match based on product_id, best_before_date, purchased_date and price.

If no stock row was found then a new row would be created with a new stock id.
If rows match then it checks for location and updates the amount at the location or creates a new row with the same stock_id if no location is found.

 
I pulled out the extra location picker but I am not sure how to adjust the public/views/tranfer.js to get the values correctly

**Comment by berrnd** [comment]
Thanks again! I will just try it on an example and then integrate it - maybe the coming weekend.

Haven't checked in detail, but for a transfer the `stock_id` could never be "not found" as the product is in stock, so was added via a purchase or an inventory correction booking before. I just see that id as a chain, unique per "stock add action" - so I also would just generate a new id if I buy the same product for the same price and with the same best before date twice on the same day, because that are 2 "stock add transactions" (but doesn't make any difference for now anyway...).

**Comment by kriddles** [comment]
I remember now why I just made the locationpicker2 after taking a look at the public/viewjs/components/locationpicker.js. I am not sure what to update to get it to work using a provided id instead.

**Comment by kriddles** [comment]
New commit to review
The From location will now only populate based on locations with stock.
The To location will be invalid if picked the same as the From location
+Various bugfixes

**Comment by kriddles** [comment]
+Transfer on Stock Overview page

**Comment by kriddles** [comment]
+updating amount max validation based on specific stock selection

**Comment by kriddles** [comment]
Thank you for reviewing the changes. 
I can fixup the commits so that the changes are easier to read through.

I agree with your list and getting this feature fully ready before release.

Yes, I initially set the transfer transaction as not able to be undone for the proof of concept due to the issue you described.

One question in your list-
"When "Use specific stock item" the amount field should be disabled (because the amount of the specific stock entry is used then, like on the consume page)
 And maybe use the same text like on the consume page to identify a specific stock entry"

Notice that I updated the use specific stock item dropdown to not split up a stock item and then use a default quantity of 1. Instead this list is the full stock qty which can then transfer 1 or up to max quantity that the stock holds which is why the Amount field is still required.

**Comment by berrnd** [comment]
> Notice that I updated the use specific stock item dropdown to not split up a stock item and then use a default quantity of 1. Instead this list is the full stock qty which can then transfer 1 or up to max quantity that the stock holds which is why the Amount field is still required.

Ah ok, that makes even more sense :) - we should then use this on the consume page too to keep it consistent...

**Comment by kriddles** [comment]
> > Notice that I updated the use specific stock item dropdown to not split up a stock item and then use a default quantity of 1. Instead this list is the full stock qty which can then transfer 1 or up to max quantity that the stock holds which is why the Amount field is still required.
> 
> Ah ok, that makes even more sense :) - we should then use this on the consume page too to keep it consistent...

Yes, I'll update my consume branch request to include the logic I figured out with this branch to update the "max" attribute of the amount field and see if there were other changes needed.

**Comment by kriddles** [comment]
I'm not able to reproduce your stock_id lifecyle issue you mentioned. My stock_id remains the same when I purchase one unit and transfer it multiple times.

image

**Comment by kriddles** [comment]
I fixup'd into the Transfer Products commit the changes we have made so far.

My WIP branch now contains up dates to the "transfer page" todo items on your list and some start to handling when location_id = null.

**Comment by kriddles** [comment]
WIP Branch has these items completed ready for review

> Transfer page:
> 
> * [ X]  After selecting a product, if there is only 1 location, preselect that for "From location"
> * [ X]  The success message is "Removed 1 QU of PRODUCT from stock" - should be something like "Transfered 1 QU of PRODUCT from LOCATIONFROM to LOCATIONTO"
> * [ X]  The message when from/to location is the same should be more clear (like "This cannot be the same as "From location"" or something)
>   
>   * [ ]  Or maybe just hide the selected "From location" in "To location"
(Not needed now)
> * [ X]  Maybe highlight the products default location in the location dropdowns somehow
> 
> Journal page:
> 
> * [ X]  I think we should add the location as a column there
> * [ X]  Maybe split transfers into 2 booking types "Tranfer from" and "Transfer to" to make clear from where to where the product was moved (or make this clear in a different way)
> 
> General:
> 
> * [ ]  Show all locations (and maybe the amount per location) on the product card and stock overview page, idea for this: A separate page which just shows all stock entries of a product -> load this page in a dialog when clicking the location on the product card and maybe from a new context menu entry on the stock overview page

I think the product card and stock overview page should be a separate project to work through.


> * [ ]  Undoing a transfer seems not to be possible yet (from the success message on the tranfer page and from the journal page, on the journal page this is maybe a problem as there then two rows per transaction - maybe we should introduce a `transaction_id` to identify corresponding `stock_log` rows and use then that id for undoing bookings in general)
> 
> API/Backend:
> 
> * [ ]  Treating `stock_id` as a "product lifecycle id" (as mentioned before) does not work, I purchased a product and transfered it multiple times and `stock_id` looks like this:
>   DB_Browser_for_SQLCipher_2019-11-01_19-51-05

Please review, I was unable to recreate this stock_id issue.


> * [ ]  Maybe instead of a new endpoint `/stock/products/{productId}/locations`, just add the location to the response of `/stock/products/{productId}/entries`
> 
> Demo data:
> 
> * [X ]  Seems that all demo purchases are added without a location - so tranfering any of the default demo products is not possible -> add all demo products with their default location (should be no problem for production installs, as this could only happen when `FEATURE_FLAG_STOCK_LOCATION_TRACKING` was disabled and then enabled some time later)

**Comment by berrnd** [comment]
Great, looks good to me so far... :)

Tested it now also with partial stock units (1.5 units in stock), and for that the valdiation doesn't seem to work (seems that `max` of the amount picker is always the integral number without the decimal part).

About the `stock_id` problem, had this again, steps to reproduce:
- Create a new product
- Purchase 1
- Transfer it

Must be anywhere in StockService->AddProduct.
This needs also more tests when the product has currently multiple stock lines, each from a different pruchase and therefore with a different stock id - I still think the whole removing from the old location/adding to the new location needs to be done in one loop in the TransferProduct function, as otherwise not exactly the same rows/stock_ids are used...

> The TransferProduct in StockService will need to be updated to treat the null location_ids in the stock table as the default location

I think we don't need to handle that, as mentioned above, I don't think this will be a problem with production installs (otherwise also the form validation on the transfer would need to allow "no location")

Another thing I noticed:
Seems not to work currently with a product that has "Tare weight handling" enabled...
Don't know if we should just disable transfering for that products for now, as "Tare weight handling" currently also lacks that there can be only one "container weight" per product, not per location or something similar...

**Comment by berrnd** [comment]
> About the `stock_id` problem, had this again, steps to reproduce:

I attached a database here if you want to debug this with my example, can be reproduced when you do more transfers with product "prod1 normal" (product_id 27).

[grocy.zip](https://github.com/grocy/grocy/files/3800354/grocy.zip)

**Comment by kriddles** [comment]
> Great, looks good to me so far... :)
Thanks!
>
> Tested it now also with partial stock units (1.5 units in stock), and for that the valdiation doesn't seem to work (seems that `max` of the amount picker is always the integral number without the decimal part).

Fixed partial stock issue (parseFloat not parseInt)

> Another thing I noticed:
> Seems not to work currently with a product that has "Tare weight handling" enabled...
> Don't know if we should just disable transfering for that products for now, as "Tare weight handling" currently also lacks that there can be only one "container weight" per product, not per location or something similar...

Disabled for Tare weight handling for now

**Comment by kriddles** [comment]
> > About the `stock_id` problem, had this again, steps to reproduce:
> 
> I attached a database here if you want to debug this with my example, can be reproduced when you do more transfers with product "prod1 normal" (product_id 27).
> 
> [grocy.zip](https://github.com/grocy/grocy/files/3800354/grocy.zip)

Thanks, I'll take a look further with this DB. I started with a fresh DB built from migrations and did not see the issue.

I make sure to Empty Browser Cache and Hard Reload, perhaps that might clear up the bug?

I also added the Consume commits but still need to go back and update a few things with that page.

**Comment by berrnd** [comment]
I always use the default demo data for testing (just delete the database file, put an empty file `demo.txt` in the data directory and it populates a new DB by [`DemoDataGeneratorService.php`](https://github.com/grocy/grocy/blob/master/services/DemoDataGeneratorService.php)).

**Comment by kriddles** [comment]
> I always use the default demo data for testing (just delete the database file, put an empty file `demo.txt` in the data directory and it populates a new DB by [`DemoDataGeneratorService.php`](https://github.com/grocy/grocy/blob/master/services/DemoDataGeneratorService.php)).

I fixed a bug in the AddProduct which caused the issue you were seeing with the stock_id lifecyle.

I updated the consume logic to reflect the transfer logic using location information which is now in the first commit.

This is ready for another major review/testing. I fixedup into the Transfer Products branch all of the Transfer feature logic.

File: services/StockService.php [hunk 1]
```diff
@@ -214,43 +238,196 @@ public function AddProduct(int $productId, float $amount, $bestBeforeDate, $tran
...
+
+				$stockEntry->update(array(
+					'amount' => $stockEntry->amount - $amount
+				));
+				$amount = 0;
+			}
+		}
+	}
+
 	public function ConsumeProduct(int $productId, float $amount, bool $spoiled, $transactionType, $specificStockEntryId = 'default', $recipeId = null)
```
Comment (berrnd) [inline services/StockService.php]: When it's now possible to consume a product from a specific location, how is this ensured when here no `location_id` is considered... (?)

**Comment by kriddles** [comment]
> General comments:
> 
> * I think it's not worth it to do all the "location_id null checking" - as mentionend before, I think this cannot happen on production installs, is only a problem of the current demo data, but it would be just easier to ensure that the `location_id` is never `null`, for example by a SQL trigger or something which just sets the products default location if not supplied

Thank you for reviewing. 
I'll take a step back on these updates for now while you give it a go for a round for edits.

File: services/StockService.php (hunk 1 above)
Comment (kriddles) [reply services/StockService.php]: Ha, good point... I think I initially looked at it for specific stock entry and did not test without using that feature.

File: public/viewjs/consume.js [hunk 2]
```diff
@@ -178,13 +181,67 @@ $('#save-mark-as-open-button').on('click', function(e)
...
+                        function(stockEntries)
+                        {
+                                stockEntries.forEach(stockEntry =>
+                                {
+                                        if (stockEntry.location_id == locationId)
+                                        {
+                                                $("#specific_stock_entry").append($("<option>", {
+                                                        value: stockEntry.stock_id,
+                                                        amount: stockEntry.amount,
+                                                        text: __t("Amount remaining: %1$s, Best Before: %2$s", stockEntry.amount, stockEntry.best_before_date)
```
[the remark this answers is not in the document]
Comment (kriddles) [reply public/viewjs/consume.js]: Yes, and perhaps also the price?

File: services/StockService.php [hunk 3]
```diff
@@ -214,43 +238,196 @@ public function AddProduct(int $productId, float $amount, $bestBeforeDate, $tran
...
-				'best_before_date' => $bestBeforeDate,
-				'purchased_date' => $purchasedDate,
-				'stock_id' => $stockId,
-				'transaction_type' => $transactionType,
-				'price' => $price,
-				'location_id' => $locationId
-			));
-			$logRow->save();
+			// Create a new stock_id
+			if (empty($stockRows))
```
[the remark this answers is not in the document]
Comment (kriddles) [reply services/StockService.php]: stockRows checks against product_id, best_before_date, purchased_date and price. It should only create a new stock_id if there is not stock that matches with those 4 values. 
The following elseif checks against a stockRow which should limit stockRows based on location. It then will add based on the available stock_id. 
If there is stock at the location the next block should catch that and update the amount at that location.

**Comment by kriddles** [comment]
-rebased to your master

**Comment by berrnd** [comment]
> I'll take a step back on these updates for now while you give it a go for a round for edits.

Ok, I will fiddle something in the next days...

File: public/viewjs/consume.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (berrnd) [reply public/viewjs/consume.js]: Sure, but only if `FEATURE_FLAG_STOCK_PRICE_TRACKING` is enabled...

**Comment by kriddles** [comment]
This is great! Thank you. 
I'll play around with it some more but initial testing looks good.

**Comment by berrnd** [comment]
> - [ ] Undoing transfer bookings, incl. introducing a transaction_id in general for corresponding bookings and use only that transaction_id to undo bookings/actions

I started the database part how I think this could work:
- A new column `correlation_id` for `stock_log`: This groups dependent log entries together (currently only for `transfer_from`/`transfer_to` - means when one of the bookings is undone, the other one(s) have to be undone as well
- A new column `transaction_id` for `stock_log`: This groups together all affected stock rows of a action, example (default demo data): 5 Packs cheese, each bought through a different purchase -> means 5 single rows with an amount of 1 each in the `stock` table with different `stock_id`s, all 5 are transfered at once => will result in one unique `transaction_id`, 5 `correlation_id`s

grafik

TODO:
- [ ] Return the `transaction_id` in addition to the current `booking_id` via the API (to not break it)
- [ ] Undo by `transaction_id` on purchase/consume/transfer/inventory
- [ ] Undo dependent bookings by `correlation_id`

**Comment by kriddles** [comment]
Taking a look at the Stock Overview page, not sure if it was what you were thinking but an initial branch is there to look at.

I added price, location, and purchased date columns along with loading the Stock table instead of the overview. I added a checkbox with no functionality yet which I think would be nice if you could switch views. I've not yet figured out how to do that yet.

**Comment by berrnd** [comment]
Hm, if the views can be switched, maybe ok - I would not add too much columns to the stock overview table (or make it at least configurable then which columns to show)...

If there are two tables, on any action the corresponding other row(s) would need to be updated also - maybe a lot to do without always reloading the page...

So I think the easiest thing would be just to show the detail stock entries (like you added it) on a separate page and load this page in a dialog.

For the product card, I can also think of that all locations with the corresponding amount are just rendered in a tooltip when hovering the location.

Another idea for the stock overivew table: Child rows to show the stock entries when expanding a product row (https://datatables.net/examples/api/row_details.html).

**Comment by kriddles** [comment]
A little more progress... 
New stock detail page based on overview page - (needs an icon)

Instead of updating child rows which does look cool, I went with a new page mainly because it's an easy jump to making stock edits which I know was a request. Initial wip branch on stock edit page available initially in the stock detail stock selection dropdown

I've not begun much working through the stockedit.js logic yet

image

**Comment by kriddles** [comment]
Fixed detail to make the Consume 1 button work specific to that stock entry and location.
After using this button, how do you refresh the row/datatable appropriately?

**Comment by berrnd** [comment]
Cool, lets finish this with that approach:

So all `stock` table fields should be editable on the "stockedit" page (of course not including internal fields like `id`, `stock_id` and so on).
I think one `stock_log` entry with the edited values is enough.

> There could be further advanced logic that would update any historical stock_log rows based on for example updating the price for a stock.

That would break that `stock_log` is a log of what happened with a stock entry... :D

I would only show the table on the "stockdetail" page (without the header stuff) and load it in a dialog from a context menu entry on the stockoverview page pre-filtered for the current product (don't think that the "stockdetail" page is a good "dashboard" page, as more than one line per product gives no good overview and should only be needed if you want to look into details or need to edit something).

For the productcard, I would render all locations and the amounts (one per line) in an tooltip when hovering the default location, clicking the location would load the same pre-filtered "stockdetail" page.

If you go on with the UI stuff, I can take care about the additions for undoing bookings (undo by `transaction_id`, undo correlated entries automatically, and so on).

If this is ready, this would also close #313 - great! :)

**Comment by berrnd** [comment]
Means also, the `/stockedit` page should be opened for a stock entry, as exactly 1 entry gets edited, not a product (currently `/stockedit?embedded&product=13`).

**Comment by kriddles** [comment]
I had to head out of town for a bit. More progress on the Stock Detail. 
It's now going to be only available from the Stock Overview as requested (still in the left menu for dev). The Stock Detail will include a product picker which now has a custom search based on product id for the datatable.

Still WIP on the stock edit piece

**Comment by berrnd** [comment]
Great! :)
Also had no time the last two weeks, but had my fingers just now on the "backend" stuff:
- `StockService->UndoBooking` now also undoes all correlated bookings by `correlation_id`
- New API endpoint `/stock/transactions/{transactionId}` to list all `stock_log` entries by `transaction_id`
- New API endpoint `/stock/transactions/{transactionId}/undo` to undo multiple `stock_log` entries by `transaction_id`
- The undo button in the success message on the stockoverview/purchase/consume/inventory page now uses that new `/stock/transactions/{transactionId}/undo` endpoint

**Comment by kriddles** [comment]
> Great! :)
> Also had no time the last two weeks, but had my fingers just now on the "backend" stuff:
> 
> * `StockService->UndoBooking` now also undoes all correlated bookings by `correlation_id`
> * New API endpoint `/stock/transactions/{transactionId}` to list all `stock_log` entries by `transaction_id`
> * New API endpoint `/stock/transactions/{transactionId}/undo` to undo multiple `stock_log` entries by `transaction_id`
> * The undo button in the success message on the stockoverview/purchase/consume/inventory page now uses that new `/stock/transactions/{transactionId}/undo` endpoint

Awesome, thanks I'll pull it in and take a look too.

Now that the stock details is a popup out of the stock overview row, I don't like a popup within a popup if you go to stock edit from a stock detail.
I'll tinker with it a bit to see if I can prevent that user experience.

**Comment by kriddles** [comment]
Making some progress... but slow. 
just checking in to say hi

**Comment by kriddles** [comment]
Hi, I hope you're doing well.

I made some progress and rebased to fixup the stock detail into one commit: Initial Stock Detail Page

To prevent having multiple bootbox popups, I changed it up a bit. Would you take a look when you have a moment and provide feedback?

Now, on the stock overview page, you can select to view "Stock Details" as a dropdown option for a product. This takes you to the Stock Details page from which a dropdown selection is "Stock Edit".
The Stock Details page has a product picker included so that you can change selection on that page.

Right now the stock edit is functional but doing a direct REST PUT call on the stock table for the selected stock row so there is not currently a stock log row or undo functionality.

Currently I've only enabled editing Location, Amount, Best Before, and Price.
New date pickers would be needed to update the purchased date and opened date which I would need some help with.

I have not updated on the Stock Detail page the Consume and Transfer dropdown to be stock specific entry yet but I did update the Consume 1 and Consume All to be specific to the stock row. The refresh looks to be working as well. 

Still todo

- [ ] Consume to prefill stock specific entry 
- [ ] Transfer to prefill stock specific entry 
- [ ] Consume 1 as spoiled
- [ ] use stock edit endpoint instead of direct PUT statement

**Comment by kriddles** [comment]
Hi,

Lots of updates
- Transfer Products now allows Undo using your transaction id and correlation ids in stock_log
- Stock Detail and Edit feature is more complete including Undo functionality using stock_log
- Fixes #435 price step issues

Let me know if you see any specific bugs or issues otherwise I'd like to handoff to you to determine if you want to make underlying infrastructure changes to how I coded anything or make improvements upon the features.

**Comment by kriddles** [comment]
Rebased to your master and fixed merge conflict

**Comment by berrnd** [comment]
I made some tests through the UI und checked also the results in the database and everything looks good to me so far. Maybe the stock edit page need some more detail thoughts, some things which came up my mind:
- Maybe focus more on the stock entry edit capabilities there, put the edit button in the first row per line
- Not for all actions the corresponding row is updated (e. g. after undoing a booking or using the "open 1" button
- Disable all workflows for the productpicker there

But this and also the code can be improved later, it works, and that's my main focus in general for this project. I'm sure for some detail actions bugs will be found here and there, but that's the same with all new features I think... :)

I'll merge this now.

A real big THANK YOU for all your work on this and also your other pull request, that was by far the biggest contribution to grocy so far (not including other external tools or integrations), I really appreciate it.

**Comment by kriddles** [comment]
> I made some tests through the UI und checked also the results in the database and everything looks good to me so far. Maybe the stock edit page need some more detail thoughts, some things which came up my mind:
> 
> * Maybe focus more on the stock entry edit capabilities there, put the edit button in the first row per line
> * Not for all actions the corresponding row is updated (e. g. after undoing a booking or using the "open 1" button
> * Disable all workflows for the productpicker there
> 
> But this and also the code can be improved later, it works, and that's my main focus in general for this project. I'm sure for some detail actions bugs will be found here and there, but that's the same with all new features I think... :)
> 
> I'll merge this now.
> 
> A real big THANK YOU for all your work on this and also your other pull request, that was by far the biggest contribution to grocy so far (not including other external tools or integrations), I really appreciate it.

You're welcome! I really enjoyed working through it all. 
Thanks for your help and ideas and also thanks again for Grocy!

**Pull request merged**